### PR TITLE
Added validation for both activities(CreateAccountActivity, SignUpDetails) with unit tests for all fields

### PR DESCRIPTION
**Pull request by bakhodirsolieff**
Added validation for both activities(CreateAccountActivity, SignUpDetails) with unit tests for all fields

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -44,6 +44,8 @@ dependencies {
     implementation(libs.androidx.activity)
     implementation(libs.androidx.constraintlayout)
     testImplementation(libs.junit)
+    testImplementation("org.jetbrains.kotlin:kotlin-test")
+    testImplementation("org.robolectric:robolectric:4.9")
```
Comment (mtkar-dev) [inline app/build.gradle.kts]: Why did you add robolectric to the project here?
Same question about kotlin-test

File: app/src/main/java/com/example/myapplication/CreateAccountActivity.kt [hunk 2]
```diff
@@ -10,10 +10,15 @@ import android.widget.Toast
 import androidx.appcompat.app.AppCompatActivity
 
 class CreateAccountActivity : AppCompatActivity() {
+
+    private lateinit var credentialsManager: CredentialsManager
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CreateAccountActivity.kt]: Is there a reason to make it a lateinit var instead of a simple val?

File: app/src/main/java/com/example/myapplication/CreateAccountActivity.kt [hunk 3]
```diff
@@ -22,31 +27,31 @@ class CreateAccountActivity : AppCompatActivity() {
...
-            if (password.length < 6) {
-                passwordField.error = "Password must be at least 6 characters"
-                return@setOnClickListener
-            }
-
             val rememberMe = rememberMeCheckbox.isChecked
 
-            Toast.makeText(this, "Signed in", Toast.LENGTH_SHORT).show()
+            when {
+                email.isEmpty() -> emailField.error = "Email is required"
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CreateAccountActivity.kt]: 1) use string resources here also, don't hardcode them
https://developer.android.com/guide/topics/resources/string-resource

2) Since you display error in two cases, how about creating a method to do it so it's more readable?
setEmailError(...)

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    private val emailPattern = ("[a-zA-Z0-9\\+\\%\\-\\+]{1,256}" +
+            "\\@" +
+            "[a-zA-Z0-9][0-zA-Z0-9\\-]{0,64}" +
+            "(" +
+            "\\." +
+            "[a-zA-Z0-9][0-zA-Z0-9\\-]{0,25}" +
+            ")+").toRegex()
+
+    fun isValidEmail(email: String): Boolean {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CredentialsManager.kt]: isEmailValid seems more human for me, but since you keep it consistent - it's your preference

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        return email.matches(emailPattern)
+    }
+
+    fun isValidPassword(password: String): Boolean {
+        return password.length >= 8
+    }
+
+    fun validateCredentials(email: String, password: String, isCheckboxChecked: Boolean): Boolean {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CredentialsManager.kt:21]: what is isCheckboxChecked?

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 6]
```diff
@@ -10,9 +10,13 @@ import android.widget.Toast
 import androidx.appcompat.app.AppCompatActivity
 
 class SignUpDetailsActivity : AppCompatActivity() {
+
+    private lateinit var credentialsManager: CredentialsManager
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt:14]: Same as previously, do you need a lateinit here?

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 7]
```diff
@@ -21,24 +25,65 @@ class SignUpDetailsActivity : AppCompatActivity() {
...
+            val password = strongPasswordEditText.text.toString().trim()
             val isTermsAccepted = termsAndConditionsCheckBox.isChecked
 
-            if (isTermsAccepted) {
-                Toast.makeText(this, "You have successfully registered!", Toast.LENGTH_SHORT).show()
-            } else {
-                Toast.makeText(this, "Please accept the Terms and Conditions", Toast.LENGTH_SHORT)
-                    .show()
+            when {
+                !credentialsManager.isValidFullName(fullName) -> {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt:35]: Same as above, how about making methods for displaying errors and only passing a string resource to it?

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 8]
```diff
@@ -21,24 +25,65 @@ class SignUpDetailsActivity : AppCompatActivity() {
...
+                else -> {
+                    if (credentialsManager.ValidateCredentialsForSignUp(
+                            fullName,
+                            email,
+                            phone,
+                            password,
+                            isTermsAccepted
+                        )
+                    ) {
+                        Toast.makeText(
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt]: You have an if with both bodies doing 'exactly' the same with difference of displayed string
This is a good moment to export it to a method to avoid duplication

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        // Valid email, password, and checkbox checked
+        assertTrue(credentialsManager.validateCredentials("example@test.com", "password123", true))
+
+        // Invalid email
+        assertFalse(credentialsManager.validateCredentials("invalid-email", "password123", true))
+
+        // Invalid password
+        assertFalse(credentialsManager.validateCredentials("example@test.com", "short", true))
+
+        // Checkbox not checked
```
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt:32]: But what is a checkbox?

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            credentialsManager.validateCredentials(
+                "example@test.com",
+                "password123",
+                false
+            )
+        )
+    }
+
+    @Test
+    fun testValidFullName() {
```
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt:43]: What about one word in tested string, like "John"?

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    fun testValidPhoneNumber() {
+        assertTrue(credentialsManager.isValidPhoneNumber("1234567890"))
+        assertFalse(credentialsManager.isValidPhoneNumber("12345"))
+        assertFalse(credentialsManager.isValidPhoneNumber(""))
+    }
+
+    @Test
+    fun testTermsAccepted() {
+        assertTrue(credentialsManager.isTermsAccepted(true))
```
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt]: Making a method to test if true is true is a bit too much, is there really a need for it?

**Review by mtkar-dev — CHANGES_REQUESTED** [review]

File: app/build.gradle.kts (hunk 1 above)
Comment (bakhodirsolieff) [reply app/build.gradle.kts]: yes i will delete them, i just had some issue with running tests, and tried out many ways and eventually i fixed that with installing a new gradle

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt (hunk 5 above)
Comment (bakhodirsolieff) [reply app/src/main/java/com/example/myapplication/CredentialsManager.kt:21]: It checks whether the user ticked the checkbox before pressing the button " Next > "
